### PR TITLE
Task3 Create WeatherСonsoleApp

### PR DESCRIPTION
**Pull request by alesyalis**

### CODE REVIEW

**Review by Xodoc — APPROVED** [review]

**Review by ihhyy — APPROVED** [review]

File: Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherServices.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Weather.BL.Services.Abstract
+{
+    public interface IWeatherServices
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherServices.cs]: IWeatherService

File: Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherServices.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+    {
+        void GetWeather(string cityName);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/Abstract/IWeatherServices.cs]: Make all methods async please

File: Weather.СonsoleApp/Weather.BL/Services/WeatherServices.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Weather.BL.Services
+{
+    public class WeatherServicesv : IWeatherServices
+    {
+        public void GetWeather(string cityName)
+        {
+            try
+            {
+                string url = string.Format("http://api.openweathermap.org/data/2.5/weather?q={0}&units=metric&appid=8e943ed8b016561c73b8a1920366ef79", cityName);
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherServices.cs]: Use var

Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherServices.cs]: Move url and secret to the config file

File: Weather.СonsoleApp/Weather.BL/Services/WeatherServices.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                string response;
+
+                using (StreamReader streamReader = new StreamReader(httpWebResponse.GetResponseStream()))
+                {
+                    response = streamReader.ReadToEnd();
+                }
+
+                WeatherResponse weatherResponse = JsonConvert.DeserializeObject<WeatherResponse>(response);
+
+                Console.WriteLine("В {0}: {1} °C {2} ", weatherResponse.Name, weatherResponse.Main.Temp,
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherServices.cs]: The above logic is DAL, below is  BL

File: Weather.СonsoleApp/Weather.BL/Validators/WeatherValidator.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Weather.BL.Validators
+{
+    public class WeatherValidator
+    {
+        public static bool IsValidCityName(string cityName)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Validators/WeatherValidator.cs]: Why static

File: Weather.СonsoleApp/Weather.СonsoleApp/Program.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        static void Main(string[] args)
+        {
+            var weatherServices = new WeatherServicesv();
+
+            while (true)
+            {
+                Console.WriteLine("Enter the name of the city");
+                string city = Console.ReadLine();
+
+                if (WeatherValidator.IsValidCityName(city))
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.СonsoleApp/Program.cs]: this is BL logic

**Review by denja2 — CHANGES_REQUESTED** [review]

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+using Weather.BL.Validators.Abstract;
+using Weather.DataAccess.Models;
+using Weather.DataAccess.Repositories.Abstrdact;
+
+namespace Weather.BL.Services
+{
+    public class WeatherService : IWeatherService
+    {
+        private readonly IWeatherRepository _weatherRepository;
+        private readonly IValidator _validator;
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs:14]: Empty line

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            _validator = validator; 
+        }
+        public async Task<WeatherResponseDTO> GetWeatherAsync(string cityName)
+        {
+            _validator.ValidateCityByNameName(cityName);
+            var weather = await _weatherRepository.GetWeatherAsync(cityName);
+            
+            return MapWeather(weather);
+        }
+        private WeatherResponseDTO MapWeather(WeatherResponse weatherResponse)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: MapTo

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            var weatherDTO = new WeatherResponseDTO
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: Check for null

File: Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            if (temp < 0)
+                return "Dress warmly.";
+            if (temp >= 0 && temp <= 20)
+                return "It's fresh.";
+            if (temp > 20 && temp <= 30)
+                return "Good weather!";
+            if (temp > 30)
+                return "It's time to go to the beach";
+            else
+                return "Nothing";
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Services/WeatherService.cs]: ?

File: Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace Weather.BL.Validators.Abstract
+{
+    public class Validator: IValidator
+    {
+        public void ValidateCityByNameName(string name)
+        {
+            if (name == string.Empty)
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs]: IsNullOrWhiteSpace

File: Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+            {
+                throw new ArgumentNullException("Entering the city is required\n");
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.BL/Validators/Abstract/Validator.cs]: Create your custom ValidationException

File: Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        public WeatherRepository(IConfiguration configuration)
+        {
+            _httpClient = new HttpClient();
+            _configuration = configuration; 
+        }
+        public async Task<WeatherResponse> GetWeatherAsync(string cityName)
+        {
+            var api = _configuration.Api;
+            var key = _configuration.ApiKey;
+            var responce = await _httpClient.GetStringAsync($"{api}={cityName}&units=metric&appid={key}");
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.DataAccess/Repositories/WeatherRepository.cs]: Move the string to the config

File: Weather.СonsoleApp/Weather.СonsoleApp/Program.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            NinjectModule serviceModule = new RegistrationModule();
+
+            var kernel = new StandardKernel(serviceModule);
+            kernel.Load(Assembly.GetExecutingAssembly());
+            DependencyResolver.SetResolver(new NinjectDependencyResolver(kernel));
+            _weatherService = kernel.Get<IWeatherService>();
+            _configuration = kernel.Get<IConfiguration>();
+            
+            _configuration.Api = ConfigurationManager.AppSettings["api"];
```
Comment (denja2) [inline Weather.СonsoleApp/Weather.СonsoleApp/Program.cs]: Move to container initialization

**Review by denja2 — CHANGES_REQUESTED** [review]

**Pull request merged**